frontend/lib/resolution/scope-queries.cpp

```diff
@@ -2244,6 +2262,306 @@ void resolveUsesAndImportsInScope(Context* context, const Scope* scope) {
+             int& nOther) {
+
+  // use a set to avoid reporting a multiply-defined error
+  // for a symbol available in two ways.
+  // TODO: adjust scope resolver to not create such patterns.
+  std::set<ID> countedIds;
+
+  nParenfulMethods = 0;
+  nParenfulNonMethodFunctions = 0;
+  nParenlessMethods = 0;
+  nParenlessNonMethodFunctions = 0;
+  nFields = 0;
+  nOther = 0;
+  for (const auto& b : v) {
+    auto end = b.end();
+    for (auto it = b.begin(); it != end; ++it) {
+      const IdAndVis& idv = it.curIdAndVis();
+      auto p = countedIds.insert(idv.id());
+      if (p.second) {
+        // inserted in to the map
+        if (idv.isParenfulFunction()) {
+          if (idv.isMethodOrField()) {
+            nParenfulMethods++;
+          } else {
+            nParenfulNonMethodFunctions++;
+          }
+        } else {
+          // it is parenless or a non-function, but which of the 3 categories?
+          if (parsing::idIsFunction(context, idv.id())) {
+            if (idv.isMethodOrField()) {
+              // it must be a parenless method
+              nParenlessMethods++;
+            } else {
+              // it must be a parenless non-method function
+              nParenlessNonMethodFunctions++;
+            }
+          } else {
+            // it is not a function
+            if (idv.isMethodOrField()) {
+              nFields++;
+            } else {
+              nOther++;
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+static void countReturnIntents(Context* context,
+                               const std::vector<BorrowedIdsWithName>& v,
+                               int& nValue,
+                               int& nConstRef,
+                               int& nRef,
+                               int& nOther) {
+  nValue = 0;
+  nConstRef = 0;
+  nRef = 0;
+  nOther = 0;
+  for (const auto& b : v) {
+    for (const auto& id : b) {
+      Function::ReturnIntent reti = parsing::idToFnReturnIntent(context, id);
+      switch (reti) {
+        case Function::DEFAULT_RETURN_INTENT:
+        case Function::CONST:
+          nValue++;
+          break;
+        case Function::CONST_REF:
+          nConstRef++;
+          break;
+        case Function::REF:
+          nRef++;
+          break;
+        case Function::PARAM:
+        case Function::TYPE:
+          nOther++;
+          break;
+      }
+    }
+  }
+}
+
+
+static const bool&
+emitMultipleDefinedSymbolErrorsQuery(Context* context, const Scope* scope) {
+  QUERY_BEGIN(emitMultipleDefinedSymbolErrorsQuery, context, scope);
+
+  bool result = false;
+
+  std::set<UniqueString> namesDefined;
+  std::set<UniqueString> namesDefinedMultiply;
+  ScopeSet checkedScopes;
+
+  collectAllNames(context, scope, /* skip private */ false,
+                  namesDefined, namesDefinedMultiply, checkedScopes);
+
+  // Now, consider names in namesDefinedMultiply. If there are any
+  // that are not only parenful functions, issue an error.
```

Comment:
It would be more accurate to say something like "any that are not potentially overloaded functions" as the logic below does allow multiple parenless functions with the same name as long as they are valid return intent overloads.